File: ci/scripts/embed_json.py
import sys
import os
import json

if len(sys.argv) != 3:
    print(f"Usage: {sys.argv[0]} <input.json> <output.c>")
    sys.exit(1)

input_path = sys.argv[1]
output_path = sys.argv[2]

with open(input_path, 'r', encoding='utf-8') as f:
    data = f.read()

# Optionally, minify JSON to save space
try:
    minified = json.dumps(json.loads(data), separators=(",", ":"))
except Exception:
    minified = data

c_array = ','.join(str(ord(c)) for c in minified)

header = "#include \"diagnostic_schema_embed.h\"\n\n"
array_decl = f"const char ZT_DIAGNOSTIC_SCHEMA_JSON[] = \"{minified.replace('\\', '\\\\').replace('"', '\\"').replace(chr(10), '\\n').replace(chr(13), '')}\";\n"
len_decl = f"const unsigned int ZT_DIAGNOSTIC_SCHEMA_JSON_LEN = sizeof(ZT_DIAGNOSTIC_SCHEMA_JSON) - 1;\n"

with open(output_path, 'w', encoding='utf-8') as out:
    out.write(header)
    out.write(array_decl)
    out.write(len_decl) 

File: ci/scripts/embed_json.rs
use std::env;
use std::fs;
use std::io::Write;
use std::path::Path;

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: {} <input.json> <output.c>", args[0]);
        std::process::exit(1);
    }
    let input_path = &args[1];
    let output_path = &args[2];

    let data = fs::read_to_string(input_path).expect("Failed to read input file");
    // Minify JSON
    let minified = match serde_json::from_str::<serde_json::Value>(&data) {
        Ok(json) => serde_json::to_string(&json).unwrap_or(data.clone()),
        Err(_) => data.clone(),
    };

    let escaped = minified
        .replace('\\', "\\\\")
        .replace('"', "\\\"")
        .replace('\n', "\\n")
        .replace('\r', "");

    let header = "#include \"diagnostic_schema_embed.h\"\n\n";
    let array_decl = format!(
        "const char ZT_DIAGNOSTIC_SCHEMA_JSON[] = \"{}\";\n",
        escaped
    );
    let len_decl = "const unsigned int ZT_DIAGNOSTIC_SCHEMA_JSON_LEN = sizeof(ZT_DIAGNOSTIC_SCHEMA_JSON) - 1;\n";

    let mut out = fs::File::create(output_path).expect("Failed to create output file");
    out.write_all(header.as_bytes()).unwrap();
    out.write_all(array_decl.as_bytes()).unwrap();
    out.write_all(len_decl.as_bytes()).unwrap();
} 